crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1465,22 +1465,35 @@ impl Account {
             .into())
         } else {
             // If the event contained sender_device_keys, check them now.
+            // WARN: If you move or modify this check, ensure that the code bellow is still
+            // valid. The processing of the historic room key bundle depends on this being
+            // here.
             Self::check_sender_device_keys(event.as_ref(), sender_key)?;
 
-            // If this event is an `m.room_key` event, defer the check for the
-            // Ed25519 key of the sender until we decrypt room events. This
-            // ensures that we receive the room key even if we don't have access
-            // to the device.
-            if !matches!(*event, AnyDecryptedOlmEvent::RoomKey(_)) {
-                let Some(device) =
-                    store.get_device_from_curve_key(event.sender(), sender_key).await?
-                else {
-                    return Err(EventError::MissingSigningKey.into());
-                };
-
-                let Some(key) = device.ed25519_key() else {
-                    return Err(EventError::MissingSigningKey.into());
-                };
+            if let AnyDecryptedOlmEvent::RoomKey(_) = event.as_ref() {
+                // If this event is an `m.room_key` event, defer the check for
+                // the Ed25519 key of the sender until we decrypt room events.
+                // This ensures that we receive the room key even if we don't
+                // have access to the device.
+            } else if let AnyDecryptedOlmEvent::RoomKeyBundle(_) = event.as_ref() {
+                // If this is a room key bundle we're requiring the device keys to be part of
+                // the `AnyDecryptedOlmEvent`. This ensures that we can skip the check for the
+                // Ed25519 key bellow since `Self::check_sender_device_keys` already did so.
+                //
+                // If the event didn't contain any sender device keys we'll throw an error
+                // refusing to decrypt the room key bundle.
+                event.sender_device_keys().ok_or(EventError::MissingSigningKey).inspect_err(
```

Comment:
TIL `inspect_err` (and `inspect`)

Reply:
Yeah, they come in handy sometimes. Avoids some unnecessary branches in some simpler cases.